README
This repository contains the core parts of Voikko:

* libvoikko - The Voikko library
* voikko-fi - Finnish morphology for Voikko
* data - Shared data files used by libvoikko, Joukahainen etc.
* tests - Test data for automated tests covering the combined functionality of libvoikko and voikko-fi
* tools - Developer tools and generally useful Python libraries

The Wiki for this repository is used as the general developer Wiki for Voikko.



libvoikko/ChangeLog
This ChangeLog is no longer maintained. Please see Git history for changes
after version 4.0.2.

2016-01-30 Harri Pitkänen <hatapitk@iki.fi>
	* Look harder for spelling suggestions.

2016-01-24 Harri Pitkänen <hatapitk@iki.fi>
	* Require compiler capable of supporting the C++11 standard.

2016-01-22 Harri Pitkänen <hatapitk@iki.fi>
	* Add user application data folder to dictionary search path on Windows

2015-11-01 Harri Pitkänen <hatapitk@iki.fi>
	* Enable HFST backend by default.

2015-10-31 Harri Pitkänen <hatapitk@iki.fi>
	* Disable malaga backend by default.
	* Enable VFST backend and build tools by default.
	* Add separate option --enable-experimental-vfst for experimental
	  features related to VFST format.

2015-10-28 Harri Pitkänen <hatapitk@iki.fi>
	* Use Python 3 instead of Python 2 for build tools.

2015-10-27 Harri Pitkänen <hatapitk@iki.fi>
	* Remove default value for builtin dictionary path. No directories
	  will be searched for by default unless --with-dictionary-path
	  is specified.

2015-09-13 Harri Pitkänen <hatapitk@iki.fi>
	* Make the Python module single source compatible with Python 3.
	* Drop support for Python versions 2.5 and 2.6.
	* API methods accepting path arguments now work with both Unicode
	  strings and binary strings.

2015-08-28 Harri Pitkänen <hatapitk@iki.fi>
	* Autocorrect data is read from VFST transducer in format 5
	  dictionaries.

2015-06-20 Harri Pitkänen <hatapitk@iki.fi>
	* Add spelling suggestions to simple VFST spellers by using
	  weighted VFST transducers.

2015-06-16 Harri Pitkänen <hatapitk@iki.fi>
	* Add support for weighted VFST transducers.

2015-06-08 Harri Pitkänen <hatapitk@iki.fi>
	* Initial support for BCP 47 script subtag.

2015-03-29 Harri Pitkänen <hatapitk@iki.fi>
	* Tri-licensed the HFST backend with permission from Flammie Pirinen.

2015-01-03 Harri Pitkänen <hatapitk@iki.fi>
	* Search for dictionares from $HOME/Library/Spelling/voikko on OS X.

2014-12-29 Harri Pitkänen <hatapitk@iki.fi>
	* Add support for specifying multiple default dictionary locations.

2014-08-03 Harri Pitkänen <hatapitk@iki.fi>
	* New dictionary format 5 for VFST.

2013-12-04 Harri Pitkänen <hatapitk@iki.fi>
	* Grammar checker should allow numeric references to footnotes.

2013-11-21 Harri Pitkänen <hatapitk@iki.fi>
	* Increase version numbers for 3.7.

2013-10-10 Harri Pitkänen <hatapitk@iki.fi>
	* Add API functions for listing languages for which hypenators
	  or grammar checkers are available.

2013-10-08 Harri Pitkänen <hatapitk@iki.fi>
	* Add "voikkospell -L" for listing dictionary capabilities.

2013-09-03 Harri Pitkänen <hatapitk@iki.fi>
	* Internally standard and unspecified variant are not the same.

2013-08-21 Harri Pitkänen <hatapitk@iki.fi>
	* Improve Java API for distribution through Maven Central Repository.
	* Fix test suite to work with latest Automake releases.
	* Increase version numbers for 3.6.1.

2013-07-23 Harri Pitkänen <hatapitk@iki.fi>
	* Add option accept_all_uppercase to voikkospell.
	* Improve capitalization of suggestions.

2013-07-11 Harri Pitkänen <hatapitk@iki.fi>
	* Disable voikkovfstc by default since VFST is still experimental.

2013-04-08 Harri Pitkänen <hatapitk@iki.fi>
	* Add replacement suggestion ä->e.

2013-03-17 Harri Pitkänen <hatapitk@iki.fi>
	* Private use subtag is now always returned, default is
	  "standard". This will make it easier to deal with language
	  variants in applications.

2013-03-16 Harri Pitkänen <hatapitk@iki.fi>
	* Support for ZHFST spellers is no longer experimental.

2012-08-22 Harri Pitkänen <hatapitk@iki.fi>
	* Add support for using Python 3 (and drop support for using
	  Python < 2.6) as build tool.

2012-08-20 Harri Pitkänen <hatapitk@iki.fi>
	* Add simple VFST speller implementation.

2012-08-14 Harri Pitkänen <hatapitk@iki.fi>
	* Improve word splitting suggestions near capitalized words or
	  end of sentence.

2012-08-04 Harri Pitkänen <hatapitk@iki.fi>
	* mmap may be available even if getpwuid_r is not.
	* Add option --disable-buildtools

2012-08-03 Harri Pitkänen <hatapitk@iki.fi>
	* Support building without hardcoded DICTIONARY_PATH.

2012-08-01 Harri Pitkänen <hatapitk@iki.fi>
	* Support homonyms for verb negatives.

2012-06-18 Harri Pitkänen <hatapitk@iki.fi>
	* Fix false positives from extra verb check.
	* Update version numbers for 3.5.

2012-05-14 Harri Pitkänen <hatapitk@iki.fi>
	* Check for extra main verbs.

2012-05-01 Harri Pitkänen <hatapitk@iki.fi>
	* Codepoints in Latin Extended-B block are letters.

2012-04-01 Harri Pitkänen <hatapitk@iki.fi>
	* Update project files for .Net interface to .Net 4.0 since compiler for
	  3.5 does not seem to be the default on Debian unstable anymore.

2012-03-26 Harri Pitkänen <hatapitk@iki.fi>
	* Suggest replacement e->ä

2012-03-25 Harri Pitkänen <hatapitk@iki.fi>
	* Suggest more character replacements

2012-03-24 Harri Pitkänen <hatapitk@iki.fi>
	* Suggest some two character deletions

2012-03-21 Harri Pitkänen <hatapitk@iki.fi>
	* Generate more spelling suggestions

2012-03-06 Harri Pitkänen <hatapitk@iki.fi>
	* Suggest replacement i->e

2012-03-05 Harri Pitkänen <hatapitk@iki.fi>
	* Core code and VFST backend licensed under MPL 1.1 / GPL 2+ /
	  LGPL 2.1+ tri-license.

2012-02-27 Harri Pitkänen <hatapitk@iki.fi>
	* Add new option to voikkospell: -M for morphological analysis without
	  spell checking.

2012-02-12 Harri Pitkänen <hatapitk@iki.fi>
	* Added initial VFST lookup code.

2012-02-11 Harri Pitkänen <hatapitk@iki.fi>
	* Try replacing two characters in OCR suggestion mode.

2012-01-28 Harri Pitkänen <hatapitk@iki.fi>
	* Add canonical decompositions for Russian and Komi alphabet.

2012-01-25 Sjur Moshagen <sjurnm@mac.com>
	* Add support for cyrillic alphabet

2012-01-15 Harri Pitkänen <hatapitk@iki.fi>
	* Fix build with MSVC

2012-01-14 Harri Pitkänen <hatapitk@iki.fi>
	* Fix build with GCC 4.7
	* Update version numbers for 3.4.1

2011-12-15 Harri Pitkänen <hatapitk@iki.fi>
	* Be more careful with ordinal ranges.


tools/pylib/hfconv.py
# -*- coding: utf-8 -*-

# Copyright 2005, 2006 Harri Pitkänen (hatapitk@cc.jyu.fi)
# Functions and data for Hunspell-fi <-> Suomi-malaga converter

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import re

grads = [ (u'sw', u'tt', u'av1'),
	(u'sw', u'pp', u'av1'),
	(u'sw', u'kk', u'av1'),
	(u'sw', u'mp', u'av1'),
	(u'sw', u'p', u'av1'),
	(u'sw', u'nt', u'av1'),
	(u'sw', u'lt', u'av1'),
	(u'sw', u'rt', u'av1'),
	(u'sw', u't', u'av1'),
	(u'sw', u'nk', u'av1'),
	(u'sw', u'uku', u'av1'),
	(u'sw', u'yky', u'av1'),
	(u'ws', u't', u'av2'),
	(u'ws', u'p', u'av2'),
	(u'ws', u'k', u'av2'),
	(u'ws', u'mm', u'av2'),
	(u'ws', u'v', u'av2'),
	(u'ws', u'nn', u'av2'),
	(u'ws', u'll', u'av2'),
	(u'ws', u'rr', u'av2'),
	(u'ws', u'd', u'av2'),
	(u'ws', u'ng', u'av2'),
	(u'sw', u'k>j', u'av3'),
	(u'ws', u'j>k', u'av4'),
	(u'sw', u'k>', u'av5'),
	(u'ws', u'>k', u'av6')   ]

# Joukahainen word classes
SUBST = 1
ADJ = 2
VERB = 3

classmap = [(u'valo', u'sw', [(None,u'(.*)',u'valo'),
			(u'k>',u'(ko)ko',u'koko'),
			(u'k>',u'(ruo)ko',u'ruoko'),
			(u'kk',u'(.*k)kU',u'alku'),
			(u'uku',u'(.U)kU',u'luku'),
			(u'k>',u'(..U)kU',u'tiuku'),
			(u'k>',u'(.*)kU',u'alku'),
			(u'lt',u'(.*l)tO',u'aalto'),
			(u'nt',u'(.*n)tO',u'anto'),
			(u'nt',u'(.*n)tU',u'lintu'),
			(u'nk',u'(.*n)kO',u'hanko'),
			(u'tt',u'(.*t)tU',u'hattu'),
			(u'tt',u'(.*t)tO',u'liitto'),
			(u'nk',u'(.*n)kU',u'hinku'),
			(u'pp',u'(.*p)pU',u'hoppu'),
			(u'rt',u'(.*r)tO',u'kaarto'),
			(u'pp',u'(.*p)pO',u'kippo'),
			(u'mp',u'(.*m)pO',u'sampo'),
			(u'mp',u'(.*m)pU',u'kumpu'),
			(u't',u'(.*)tU',u'laatu'),
			(u'p',u'(.*)pU',u'apu'),
			(u'p',u'(.*)pO',u'lepo'),
			(u't',u'(.*)tO',u'leuto'),
			(u'kk',u'(.*k)kO',u'verkko'),
			(u'k>',u'(.*h)kO',u'vihko'),
			(u'k>',u'(.*)kO',u'verkko')   ]),
	(u'arvelu', u'sw', [(None,u'(.*Ce[lr])O',u'hontelo',[ADJ]),
			(None,u'(.*)',u'arvelu'),
			(u'nk',u'(.*n)kO',u'alanko'),
			(u'nt',u'(.*n)tO',u'avanto'),
			(u'kk',u'(.*k)kO',u'laatikko'),
			(u'tt',u'(.*t)tO',u'pihatto'),
			(u'tt',u'(.*t)tU',u'raamattu') ]),
	(u'autio', u'-', [(None,u'(.*)',u'autio')]),
	(u'kiiski', u'-', [(None,u'(.*)i',u'kiiski')]),
	(u'risti', u'sw', [(None,u'(.*)i',u'risti'),
			(u'pp',u'(pop)pi',u'pop'),
			(u'pp',u'(.*p)pi',u'keppi'),
			(u'lt',u'(.*l)ti',u'pelti'),
			(u'nk',u'(.*n)ki',u'renki'),
			(u'kk',u'(punk)ki',u'punk'),
			(u'kk',u'(.*k)ki',u'takki'),
			(u'tt',u'(.*t)ti',u'tatti'),
			(u'nt',u'(.*n)ti',u'tunti'),
			(u'p',u'(.*)pi',u'hupi'),
			(u't',u'(.*)ti',u'vati'),
			(u'k>',u'(.*)ki',u'takki')]),
	(u'paperi', u'sw', [(None,u'(.*)i',u'paperi'),
			(u'nt',u'(.*n)ti',u'hollanti'),
			(u'nk',u'(.*n)ki',u'killinki'),
			(u'kk',u'(.*k)ki',u'kajakki'),
			(u'tt',u'(.*t)ti',u'salaatti'),
			(u'pp',u'(.*p)pi',u'sinappi'),
			(u't',u'(.*)ti',u'konvehti') ]),
	(u'edam', u'-', [(None,u'(.*C)',u'edam')]),
	(u'kalsium', u'-', [(None,u'(.*)i',u'fan'),
			 (None,u'(.*)',u'kalsium')]),
	(u'lovi', u'sw',   [(None,u'(.*)i',u'lovi'),
			(u'nk',u'(.*n)ki',u'hanki'),
			(u'pp',u'(.*p)pi',u'happi'),
			(u'mp',u'(.*lam)pi',u'lampi'),
			(u'mp',u'(.*m)pi',u'sampi'),
			(u'kk',u'(.*k)ki',u'kaikki'),
			(u'k>j',u'(.*)ki',u'kylki'),
			(u't',u'(.*lah)ti',u'lahti'),
			(u't',u'(.*h)ti',u'lehti'),
			(u'p',u'(.*)pi',u'siipi'),
			(u'k>',u'(.*)ki',u'kaikki')]),
	(u'nalle', u'sw', [(None,u'(.*Ce)',u'nalle'),
		         (u'tt',u'(.*t)te',u'atte'),
		         (u'kk',u'(.*k)ke',u'nukke')]),
	(u'kala', u'sw',   [(None,u'(.*)A',u'kala'),
			(u'tt',u'(.*t)tA',u'aitta'),
			(u'nk',u'(.*n)kA',u'hanka'),
			(u'mp',u'(.*m)pA',u'kampa'),
			(u'nt',u'(.*n)tA',u'kanta'),
			(u'pp',u'(.*p)pA',u'kappa'),
			(u'rt',u'(.*r)tA',u'parta'),
			(u'lt',u'(.*l)tA',u'valta'),
			(u'kk',u'(.*k)kA',u'haka'),
			(u'p',u'(.*)pA',u'napa'),
			(u't',u'(.*)tA',u'pata'),
			(u'k>',u'(.*AA)kA',u'raaka'),
			(u'k>',u'(.*)kA',u'haka')]),
	(u'koira', u'sw',   [(None,u'(.*)A',u'koira'),
			(u'tt',u'(.*t)tA',u'kenttä'),
			(u'nk',u'(.*n)kA',u'honka'),
			(u'mp',u'(.*m)pA',u'kompa'),
			(u'nt',u'(.*n)tA',u'suunta'),
			(u'pp',u'(.*p)pA',u'tolppa'),
			(u'rt',u'(.*r)tA',u'turta'),
			(u'lt',u'(.*l)tA',u'kulta'),
			(u'kk',u'(.*k)kA',u'hoikka'),
			(u'p',u'(.*)pA',u'huopa'),
			(u't',u'(.*)tA',u'juhta'),
			(u'k>',u'(.*i)kA',u'ikä'),
			(u'k>',u'(.*)kA',u'hoikka')]),
	(u'matala', u'-', [(None,u'(.*C)A',u'matala')]),
	(u'asema', u'sw',  [(None,u'(.*)A',u'asema'),
			(u'nt',u'(.*n)tA',u'emäntä')]),
	(u'kulkija', u'-', [(None,u'(.*i)jA',u'kulkija'),
			(None,u'(.*)A',u'apila')]),
	(u'karahka', u'sw', [(None,u'(.*)A',u'karahka'),
			(u'tt',u'(.*t)tA',u'savotta'),
			(u'pp',u'(.*p)pA',u'ulappa'),
			(u'kk',u'(.*k)kA',u'solakka')]),
	(u'apaja', u'-', [(None,u'(.*C)A',u'apaja')]),
	(u'peruna', u'-', [(None,u'(.*C)A',u'peruna')]),
	(u'herttua', u'-', [(None,u'(.*tU)A',u'herttua')]),
	(u'korkea', u'-', [(None,u'(.*C)eA',u'korkea'),
		         (None,u'(.*O)A',u'ainoa')]),
	(u'suurempi', u'sw', [(u'mp',u'(.*V)mpi',u'suurempi')]),
	(u'vapaa', u'-', [(None,u'(.*CA)A',u'vapaa'),
		        (None,u'(.*CO)O',u'tienoo'),
		        (None,u'(.*CU)U',u'leikkuu')]),
	(u'kamee', u'-',   [(None,u'(.*Ce)e',u'kamee'),
			(None,u'(.*CO)O',u'trikoo'),
			(None,u'(.*CU)U',u'revyy')]),
	(u'pii', u'-', [(None,u'(.*V)i',u'pii'),
		      (None,u'(.*CA)A',u'maa'),
		      (None,u'(.*Ce)e',u'tee'),
		      (None,u'(.*U)U',u'puu')]),
	(u'suo', u'-', [(None,u'(.*C)UO',u'suo')]),
	(u'askel', u'ws', [(None,u'(.*VC)',u'askel'),
		         (u'nn',u'(.*n)nel',u'kannel'),
		         (u'nn',u'(.*n)ner',u'kinner'),
		         (u'nn',u'(.*n)nAr',u'piennar'),
		         (u'mm',u'(.*m)mel',u'ommel'),
		         (u'ng',u'(.*n)ger',u'penger'),
		         (u'v',u'(.*)vAl',u'taival')]),
	(u'rosé', u'-', [(None,u'(.*V)',u'rosé')]),
	(u'parfait', u'-', [(None,u'(.*VC)',u'parfait')]),
	(u'huuli', u'-', [(None,u'(.*C)i',u'tuohi')]),
	(u'tuohi', u'-', [(None,u'(.*C)i',u'lohi')]),
	(u'niemi', u'-', [(None,u'(.*V)mi',u'niemi')]),
	(u'lumi', u'-', [(None,u'(.*V)mi',u'lumi')]),
	(u'susi', u'-', [(None,u'(.*V)si',u'susi')]),
	(u'kansi', u'-', [(None,u'(.*n)si',u'kansi'),
	                  (None,u'(.*r)si',u'hirsi'),
		        (None,u'(.*l)si',u'jälsi')]),
	(u'sisar', u'ws', [(None,u'(.*CVC)',u'sisar'),
		         (u't',u'(.*t)Ar',u'tytär')]),
	(u'uistin', u'ws', [(None,u'(.*i)n',u'uistin'),
	                    (u'nn',u'(.*n)nin',u'vaimennin'),
		          (u'll',u'(.*l)lin',u'sivellin'),
		          (u'rr',u'(.*r)rin',u'kiharrin'),
		          (u'd',u'(.*)din',u'kaadin'),
		          (u'v',u'(.*)vin',u'kaavin'),
			(u't',u'(.*t)in',u'suodatin'),
		          (u'j>k',u'(.*l)jin',u'poljin'),
		          (u'>k',u'(.*)in',u'puin')]),
	(u'onneton', u'ws', [(None,u'(.*t)On',u'alaston'),
	                     (u't',u'(.*t)On',u'onneton')]),
	(u'sisin', u'', [(None,u'(.*)in',u'pahin')]),
	(u'nainen', u'-', [(None,u'(.*)nen',u'nainen')]),
	(u'vastaus', u'-', [(None,u'(.*V)s',u'vastaus')]),
	(u'kalleus', u'-', [(None,u'(.*VU)s',u'kalleus'),
	                    (None,u'(.*vU)s',u'kalleus')]),
	(u'vieras', u'ws', [(None,u'(.*A)s',u'vieras'),
	                    (None,u'(.*)is',u'kauris'),
			(None,u'(.*e)s',u'kirves'),
	                    (u'nn',u'(.*n)nAs',u'kinnas'),
		          (u'll',u'(.*l)lAs',u'allas'),
		          (u'rr',u'(.*r)rAs',u'harras'),
		          (u'mm',u'(.*m)mAs',u'hammas'),
		          (u'ng',u'(.*n)gAs',u'kangas'),
			(u'k',u'(.*k)As',u'avokas',[SUBST]),
			(u'k',u'(.*k)As',u'vilkas',[ADJ]),
		          (u'p',u'(.*p)As',u'saapas'),
		          (u'd',u'(.*)dAs',u'ahdas'),
		          (u'v',u'(.*)vAs',u'varvas'),
		          (u't',u'(.*t)As',u'ratas'),
		          (u'>k',u'(.*)As',u'varas'),
			(u'>k',u'(.*)is',u'ruis'),
			(u'>k',u'(.*)es',u'ies')]),
	(u'iäkäs', u'ws', [(u'k',u'(.*k)As',u'iäkäs',[ADJ]),
		         (u'k',u'(.*k)As',u'asiakas',[SUBST])]),
	(u'ohut', u'-', [(None,u'(.*CU)t',u'airut')]),
	(u'mies', u'-', [(None,u'(.*mie)s',u'mies')]),
	(u'kuollut', u'-', [(None,u'(.*C)Ut',u'kuollut')]),
	(u'hame', u'ws', [(None,u'(.*e)',u'hame'),
	                  (u'nn',u'(.*n)ne',u'enne'),
		        (u'll',u'(.*l)le',u'helle'),
		        (u'rr',u'(.*r)re',u'kierre'),
		        (u'mm',u'(.*m)me',u'lumme'),
		        (u'j>k',u'(.*C)je',u'lahje'),
		        (u'p',u'(.*p)e',u'lape'),
		        (u'd',u'(.*)de',u'sade'),
		        (u'v',u'(.*)ve',u'taive'),
		        (u'k',u'(.*k)e',u'tarvike'),
		        (u'>k',u'(.*V)e',u'tarvike'),
		        (u't',u'(.*Vt)e',u'vaate')]),
	# Verbs
	(u'punoa', u'sw', [(None,u'(.*)A',u'punoa'),
	                   (u'mp',u'(.*m)pUA',u'ampua'),
	                   (u'mp',u'(.*m)pOA',u'tempoa'),
		         (u'tt',u'(.*t)tUA',u'asettua'),
		         (u'tt',u'(.*t)tOA',u'viittoa'),
		         (u'kk',u'(.*k)kOA',u'aikoa'),
		         (u'kk',u'(.*k)kUA',u'kiekua'),
		         (u'pp',u'(.*p)pOA',u'harppoa'),
		         (u'pp',u'(.*p)pUA',u'kieppua'),
		         (u'nt',u'(.*n)tUA',u'jakaantua'),
		         (u'rt',u'(.*r)tOA',u'kertoa'),
		         (u'rt',u'(.*r)tUA',u'kumartua'),
		         (u'nk',u'(.*n)kUA',u'mankua'),
		         (u'nk',u'(.*n)kOA',u'penkoa'),
		         (u'lt',u'(.*l)tUA',u'paleltua'),
		         (u't',u'(.*)tUA',u'kaatua'),
		         (u't',u'(.*)tOA',u'tahtoa'),
		         (u'p',u'(.*)pOA',u'leipoa'),
		         (u'p',u'(.*)pUA',u'saapua'),
		         (u'k>',u'(.*U)kUA',u'liukua'),
		         (u'k>',u'(.*)kOA',u'aikoa'),
		         (u'k>',u'(.*)kUA',u'kiekua')]),
	(u'aavistaa', u'sw', [(None,u'(.*t)AA',u'aavistaa'),
	                      (u'tt',u'(.*t)tAA',u'alittaa'),
			  (u't',u'(.*h)tAA',u'astahtaa')]),
	(u'hidastaa', u'-', [(None,u'(.*t)AA',u'hidastaa')]),
	(u'heittää', u'sw', [(u'tt',u'(.*t)tAA',u'heittää')]),
	(u'muistaa', u'-', [(None,u'(.*C)AA',u'muistaa')]),
	(u'inttää', u'sw', [(u'tt',u'(.*t)tAA',u'inttää')]),
	(u'sulaa', u'sw', [(None,u'(.*C)AA',u'sulaa'),
	                   (u'nt',u'(.*n)tAA',u'kyntää'),
		         (u'tt',u'(.*t)tAA',u'autioittaa'),
		         (u'k>',u'(.*C)kAA',u'purkaa')]),
	(u'hohtaa', u'sw', [(u't',u'(.*)tAA',u'hohtaa')]),
	(u'hujahtaa', u'sw', [(u't',u'(.*V)htAA',u'hujahtaa')]),
	(u'kirjoittaa', u'sw', [(u'tt',u'(.*V)ittAA',u'kirjoittaa'),
			    (u'tt',u'(.*V)ttAA',u'ammottaa')]),
	(u'loistaa', u'-', [(None,u'(.*C)AA',u'loistaa')]),
	(u'vuotaa', u'sw', [(u'lt',u'(.*Vl)tAA',u'puoltaa'),
			(u't',u'(.*V)tAA',u'vuotaa')]),
	(u'huutaa', u'sw', [(u'nt',u'(.*Vn)tAA',u'huutaa'),
			(u't',u'(.*V)tAA',u'huutaa')]),
	(u'sukeltaa', u'sw', [(u'lt',u'(.*Vl)tAA',u'sukeltaa'),
			  (u'rt',u'(.*Vr)tAA',u'musertaa')]),
	(u'paleltaa', u'sw', [(u'lt',u'(.*Vl)tAA',u'paleltaa')]),
	(u'murtaa', u'sw', [(u'rt',u'(.*Vr)tAA',u'murtaa')]),
	(u'juontaa', u'sw', [(u'nt',u'(.*Vn)tAA',u'juontaa')]),
	(u'pahentaa', u'sw', [(u'nt',u'(.*Vn)tAA',u'pahentaa')]),
	(u'kaivaa', u'sw', [(None,u'(.*C)AA',u'kaivaa'),
	                    (u'nt',u'(.*n)tAA',u'antaa'),
			(u'pp',u'(.*p)pAA',u'lappaa'),
			(u'tt',u'(.*t)tAA',u'saattaa'),
			(u'kk',u'(.*k)kAA',u'jakaa'),
			(u'k>',u'(.*)kAA',u'jakaa'),
			(u't',u'(.*)tAA',u'raataa')]),
	(u'kaikaa', u'-', [(None,u'(.*C)AA',u'kapsaa')]),
	(u'soutaa', u'sw', [(u't',u'(.*V)tAA',u'soutaa')]),
	(u'laskea', u'sw', [(None,u'(.*C)eA',u'laskea'),
	                    (u'nk',u'(.*n)keA',u'tunkea'),
			(u't',u'(.*)teA',u'kutea'),
			(u'kk',u'(.*k)keA',u'hakea'),
			(u'p',u'(.*)peA',u'rypeä'),
			(u'k>j',u'(.*)keA',u'polkea'),
			(u'k>',u'(.*)keA',u'hakea')]),
	(u'sallia', u'sw', [(None,u'(.*C)iA',u'sallia'),
			(u'nk',u'(.*n)kiA',u'onkia'),
			(u'mp',u'(.*m)piA',u'empiä'),
			(u'nt',u'(.*n)tiA',u'kontia'),
			(u'pp',u'(.*p)piA',u'oppia'),
			(u'kk',u'(.*k)kiA',u'loikkia'),
			(u'tt',u'(.*t)tiA',u'sättiä'),
			(u't',u'(.*)tiA',u'laatia'),
			(u'p',u'(.*)piA',u'kaapia'),
			(u'k>j',u'(.*)kiA',u'hylkiä'),
			(u'k>',u'(.*i)kiA',u'poikia'),
			(u'k>',u'(.*)kiA',u'loikkia')]),
	(u'moittia', u'sw', [(u'tt',u'(.*t)tiA',u'moittia')]),
	(u'voida', u'ws', [(u't',u'(.*)idA',u'voida')]),
	(u'kanavoida', u'ws', [(u't',u'(.*O)idA',u'kanavoida')]),
	(u'juoda', u'-', [(None,u'(.*C)UOdA',u'juoda'),
		        (None,u'(.*C)iedA',u'viedä')]),
	(u'nuolaista', u'ws', [(None,u'(CAis)tA',u'nousta'),
			   (None,u'(.*CA)istA',u'nuolaista'),
			   (None,u'(.*C)istA',u'kalista'),
			   (None,u'(.*s)tA',u'nousta'),
			   (u'ng',u'(.*n)gAistA',u'rangaista')]),
	(u'mennä', u'-', [(None,u'(.*n)nA',u'mennä')]),
	(u'katsella', u'ws', [(None,u'(.*el)lA',u'katsella'),
			(None,u'(.*eil)lA',u'katsella'),
			(None,u'(.*Vil)lA',u'arvailla'),
			(None,u'(.*il)lA',u'katsella'),
			(None,u'(.*Ol)lA',u'tulla'),
			(None,u'(.*Ul)lA',u'tulla'),
			(u'mm',u'(.*m)mellA',u'ommella'),
			(u'rr',u'(.*r)rellA',u'askarrella'),
			(u'nn',u'(.*n)nellA',u'pienennellä'),
			(u'll',u'(.*l)lellA',u'takellella'),
			(u'k',u'(.*k)ellA',u'nakella'),
			(u't',u'(.*t)ellA',u'aatella'),
			(u'p',u'(.*p)ellA',u'tapella'),
			(u'd',u'(.*)dellA',u'kohdella'),
			(u'>k',u'(.*)ellA',u'nakella')]),
	(u'arvailla', u'-', [(None,u'(.*Vi)llA',u'arvailla')]),
	(u'haravoida', u'ws', [(u't',u'(.*O)idA',u'haravoida')]),
	(u'valita', u'-', [(None,u'(.*i)tA',u'valita')]),
	(u'saneerata', u'-', [(None,u'(.*C)AtA',u'saneerata')]),
	(u'aleta', u'ws',  [(None,u'(.*CV)tA',u'aleta'),
	                    (u'mm',u'(.*m)metA',u'lämmetä'),
			(u't',u'(.*t)OtA',u'loitota'),
			(u'p',u'(.*p)AtA',u'hapata'),
			(u'p',u'(.*p)etA',u'suipeta'),
			(u'k',u'(.*k)etA',u'vaieta'),
			(u'd',u'(.*)detA',u'edetä'),
			(u'd',u'(.*)dOtA',u'leudota'),
			(u'd',u'(.*)dAtA',u'mädätä'),
			(u'v',u'(.*)vetA',u'kaveta'),
			(u'j>k',u'(.*)jetA',u'tarjeta'),
			(u'>k',u'(.*)OtA',u'ulota'),
			(u'>k',u'(.*)etA',u'vaieta')]),
	(u'haluta', u'ws', [(None,u'(.*C)itA',u'selvitä'),
			(None,u'(.*U)tA',u'haluta'),
			(u'll',u'(.*l)litA',u'hellitä'),
			(u'mm',u'(.*m)mitA',u'lämmitä'),
			(u'p',u'(.*p)UtA',u'silputa'),
			(u'v',u'(.*)vUtA',u'vivuta')]),
	(u'juoruta', u'ws', [(None,u'(.*U)tA',u'juoruta'),
			(u'mm',u'(.*m)mUtA',u'kummuta'),
			(u't',u'(.*t)UtA',u'luututa'),
			(u'p',u'(.*p)UtA',u'ryöpytä'),
			(u'v',u'(.*)vUtA',u'kavuta')]),
	(u'salata', u'ws', [(None,u'(.*)AtA',u'salata'),
	                    (u'ng',u'(.*n)gAtA',u'hangata'),
			(u'mm',u'(.*m)mAtA',u'kammata'),
			(u'rr',u'(.*r)rAtA',u'kerrata'),
			(u'nn',u'(.*n)nAtA',u'suunnata'),
			(u'll',u'(.*l)lAtA',u'vallata'),
			(u'v',u'(.*)vAtA',u'kelvata'),
			(u't',u'(.*t)AtA',u'kuitata'),
			(u'd',u'(.*)dAtA',u'ladata'),
			(u'k',u'(.*k)AtA',u'pakata'),
			(u'p',u'(.*p)AtA',u'pompata'),
			(u'>k',u'(.*)AtA',u'taata')]),
	(u'katketa', u'ws', [(None,u'(.*e)tA',u'katketa'),
	                     (u'mm',u'(.*m)metA',u'kammeta'),
			 (u'ng',u'(.*n)getA',u'langeta'),
			 (u't',u'(.*t)OtA',u'lotota'),
			 (u'k',u'(.*k)etA',u'poiketa'),
			 (u'v',u'(.*)vetA',u'ruveta'),
			 (u'd',u'(.*)detA',u'todeta'),
			 (u'j>k',u'(.*)jetA',u'lohjeta'),
			 (u'>k',u'(.*)OtA',u'saota'),
			 (u'>k',u'(.*)etA',u'poiketa')]),
	(u'kohota', u'ws', [(None,u'(.*O)tA',u'kohota'),
	                    (u'rr',u'(.*r)rOtA',u'irrota'),
			(u'mm',u'(.*m)mOtA',u'kimmota'),
			(u'ng',u'(.*n)gOtA',u'lingota'),
			(u't',u'(.*t)OtA',u'netota'),
			(u'p',u'(.*p)OtA',u'upota'),
			(u'v',u'(.*r)vOtA',u'turvota'),
			(u'k',u'(.*Vk)OtA',u'laota'),
			(u'd',u'(.*)dOtA',u'kadota'),
			(u'>k',u'(.*)OtA',u'laota')]),
	(u'kihistä', u'-', [(None,u'(.*C)istA',u'kihistä')]),
	(u'kitistä', u'-', [(None,u'(.*C)istA',u'kitistä')])
	]

flagmap = [(u'inen', 3), (u'ei_inen', 4), (u'ei_voikko', 5), (u'ei_sukija', 6), (u'ei_ys', 7),
	 (u'ei_ysa', 8), (u'ei_ysj', 9), (u'siv', 17), (u'ei_lAinen', 27),
	 (u'ysj', 29), (u'ei_mAinen', 30), (u'murre', 2), (u'yt', 31), (u'ei_vertm', 32),
	 (u'ei_yks', 37), (u'ei_voikko', 40)]

histmap = [(u'antautua', u'kaatua'), (u'banaali', u'paperi'), (u'kirjoitella', 'aatella'),
	 (u'kantaja', u'apaja'), (u'pasuuna', u'peruna'), (u'tuomi', u'niemi')]

def match_re(string, pattern):
	pattern = pattern.replace(u'V', u'(?:a|e|i|o|u|y|ä|ö|é)')
	pattern = pattern.replace(u'C', u'(?:b|c|d|f|g|h|j|k|l|m|n|p|q|r|s|t|v|w|x|y|z|š|ž)')
	pattern = pattern.replace(u'A', u'(?:a|ä)')
	pattern = pattern.replace(u'O', u'(?:o|ö)')
	pattern = pattern.replace(u'U', u'(?:u|y)')
	match = re.compile(u'^' + pattern + u'$', re.IGNORECASE).match(string)
	if match == None: return None
	else: return match.group(1)
